util/queue: add queue with fixed chunk size

wenyihu6 opened the pull request:
**util/queue: rename purge to removeAll**

This patch renames purge to removeAll since it is more descriptive of what the
function does.

Epic: none
Release note: none

---

**util/queue: add more tests**

This patch cleans the test code by modularizing the code, moving the test code
to helpers_test, and abstracting runQueueTest to use a testQueueInterface.
Currently, only Queue[T] implements it. But in a future patch, a new data
structure QueueWithFixedChunkSize[T] will also implement it.

Part of: https://github.com/cockroachdb/cockroach/issues/126560
Release note: none

--- 

**util/queue: add QueueWithFixedChunkSize**

This patch adds a new data structure QueueWithFixedChunkSize which is like Queue
but uses a fixed size for the chunked linked list. Each chunk has a fixed size
of 4000 elements. This implementation uses sync.Pool to reduce the number of
allocations. It should be used when the queue is used to store a large number of
elements.

Enqueue, Dequeue, Len, Empty run in constant time. RemoveAll runs in linear time
with respect to the number of elements in the queue. This structure is not safe
for concurrent use.

This patch adds a new data structure QueueWithFixedChunkSize which is like Queue
but uses a fixed size for the chunked linked list. Each chunk has a fixed size
of 4000 elements. This implementation uses sync.Pool to reduce the number of
allocations. It should be used when the queue is used to store a large number of
elements.

Enqueue, Dequeue, Len, Empty run in constant time. RemoveAll runs in linear time
with respect to the number of elements in the queue. This structure is not safe
for concurrent use.

```
BenchmarkQueueWithFixedChunkSize
BenchmarkQueueWithFixedChunkSize           18880            116079 ns/op               9 B/op          0 allocs/op
BenchmarkQueue
BenchmarkQueue                             28417            121969 ns/op          239060 B/op        446 allocs/op
```

Part of: https://github.com/cockroachdb/cockroach/issues/126560
Release note: none

---

**util/queue: fix linter failure**

TestLint/TestStaticCheck has some issues with generic methods. This patch works
around the issue by disabling them for these functions.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128863)
<!-- Reviewable:end -->

wenyihu6 committed:
util/queue: add more tests

This patch cleans the test code by modularizing the code, moving the test code
to helpers_test, and abstracting runQueueTest to use a testQueueInterface.
Currently, only Queue[T] implements it. But in a future patch, a new data
structure QueueWithFixedChunkSize[T] will also implement it.

Part of: #126560
Release note: none

wenyihu6 committed:
util/queue: add QueueWithFixedChunkSize

This patch adds a new data structure QueueWithFixedChunkSize which is like Queue
but uses a fixed size for the chunked linked list. Each chunk has a fixed size
of 4096 elements. This implementation uses sync.Pool to reduce the number of
allocations. It should be used when the queue is used to store a large number of
elements.

Enqueue, Dequeue, Len, Empty run in constant time. RemoveAll runs in linear time
with respect to the number of elements in the queue. This structure is not safe
for concurrent use.

```
goarch: arm64
BenchmarkQueueWithFixedChunkSize
BenchmarkQueueWithFixedChunkSize           26445            118984 ns/op              11 B/op          0 allocs/op
BenchmarkQueue
BenchmarkQueue                             28495            124318 ns/op          239060 B/op        446 allocs/op
PASS
```

Part of: https://github.com/cockroachdb/cockroach/issues/126560
Release note: none

wenyihu6 committed:
util/queue: rename purge to removeAll

This patch renames purge to removeAll since it is more descriptive of what the
function does.

Epic: none
Release note: none

nvb reviewed (commented):
> I found Oleg's implementation easier to follow

I think I do too. But I don't understand what you mean by "But I'm going with the current implementation in the Queue file". Aren't you adding `QueueWithFixedChunkSize` here?

> Reviewed 2 of 2 files at r1, 4 of 4 files at r2, 4 of 4 files at r4, 1 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-:-O4Ayznf2xzWg291J1yT:b3b52yy)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`-- commits` line 37 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4Ax8WQDC3CEiHHpGYq):*
4000 elements is a strange number. Typically factors of 2 are used for data-structures, which has some benefits like fitting nicely in heap allocation size classes. Consider switching to 4096.
___
*[`-- commits` line 48 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4B--p107BI32wPoriC):*
See my comment below about using a non-pointer type in the benchmark. These allocation numbers are hard to interpret when the queue's allocations are drowned out by the elements'.
___
*[`pkg/util/queue/queue.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4Av1pI7YBpwkyLvMk9:-O4Av1pI7YBpwkyLvMkA:b679nue) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdf073acdafc59b4abc61866e852b8f7d995e175/pkg/util/queue/queue.go#L102)):*
> ```Go
> 
> func (q *Queue[T]) removeAll() {
> 	for q.head != nil {
> ```

The following (`O(1)`) should have been enough to clear the queue and allow all chunks to be GCed. I wonder why that's not how this was written.
```go
q.head = nil
q.tail = nil
```
___
*[`pkg/util/queue/queue_with_fixed_chunk_size.go` line 65 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AyEV53A7n5GyUKgU0:-O4AyEV53A7n5GyUKgU1:b-2qvpbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/976fda2c5e8183d5d296296c09fa647c09327c38/pkg/util/queue/queue_with_fixed_chunk_size.go#L65)):*
> ```Go
> // operation would succeed.
> func (q *QueueWithFixedChunkSize[T]) Enqueue(e T) {
> 	defer func() {
> ```

Why defer this? Can't we just `q.eventCount++` immediately?
___
*[`pkg/util/queue/queue_with_fixed_chunk_size.go` line 76 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AyKOj9I_zwZRS0lYD:-O4AyKOj9I_zwZRS0lYE:b849yce) ([raw file](https://github.com/cockroachdb/cockroach/blob/976fda2c5e8183d5d296296c09fa647c09327c38/pkg/util/queue/queue_with_fixed_chunk_size.go#L76)):*
> ```Go
> 	}
> 
> 	// If the current chunk is full, create a new chunk.
> ```

This comment should make it clear that the code block below it also handles the case with a non-full chunk. It sounds like it's describing logic that would look like:
```go
if q.tail.full() {
    ...
}
```
___
*[`pkg/util/queue/queue_with_fixed_chunk_size.go` line 89 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AycJK6WHT1ZssDcXk:-O4AycJL96-hDywIhMGY:bjgv3ps) ([raw file](https://github.com/cockroachdb/cockroach/blob/976fda2c5e8183d5d296296c09fa647c09327c38/pkg/util/queue/queue_with_fixed_chunk_size.go#L89)):*
> ```Go
> // used to clear the queue properly.
> func (q *QueueWithFixedChunkSize[T]) Empty() bool {
> 	return q.head == nil || q.head.empty()
> ```

Isn't this just `q.Len() == 0`?
___
*[`pkg/util/queue/queue_with_fixed_chunk_size.go` line 117 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AzXvy1R8r6OpO2b3s:-O4AzXvzBwVewIaIAwGy:barv6h1) ([raw file](https://github.com/cockroachdb/cockroach/blob/976fda2c5e8183d5d296296c09fa647c09327c38/pkg/util/queue/queue_with_fixed_chunk_size.go#L117)):*
> ```Go
> }
> 
> // RemoveAll removes all entries in the queue and calls the callback on each
> ```

Do we need this callback? It causes the function to be less efficient and is currently required to be non-nil.
___
*[`pkg/util/queue/helpers_test.go` line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AvFOK2Yu7QtdBXmMC:-O4AvFOK2Yu7QtdBXmMD:b-5acqml) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bf73d505537ccb3170df5ea5011f3fbffbba6d/pkg/util/queue/helpers_test.go#L31)):*
> ```Go
> 	Dequeue() (*testQueueItem, bool)
> 	removeAll()
> 	checkInvariants(*testing.T, int64)
> ```

What is the int64 param?
___
*[`pkg/util/queue/helpers_test.go` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AuUJ0052UJOUcYOdm:-O4AuUJ0052UJOUcYOdn:b-t1o24g) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bf73d505537ccb3170df5ea5011f3fbffbba6d/pkg/util/queue/helpers_test.go#L35)):*
> ```Go
> }
> 
> var _ testQueueInterface = &Queue[*testQueueItem]{}
> ```

Has this file (at this commit) been run through gofmt?
___
*[`pkg/util/queue/helpers_test.go` line 78 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4Aw-l-7y2eBWi61WNV:-O4Aw-l03Gav_KcC7TLg:bcwiisg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bf73d505537ccb3170df5ea5011f3fbffbba6d/pkg/util/queue/helpers_test.go#L78)):*
> ```Go
> }
> 
> func runQueueTest(t *testing.T, q testQueueInterface, eventCount int) {
> ```

Am I missing something, or are we not using this test with `queueChunkWithFixedSize`?
___
*[`pkg/util/queue/helpers_test.go` line 78 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4Aw64t07Ea4MSD2bal:-O4Aw64t07Ea4MSD2bam:beuefzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bf73d505537ccb3170df5ea5011f3fbffbba6d/pkg/util/queue/helpers_test.go#L78)):*
> ```Go
> }
> 
> func runQueueTest(t *testing.T, q testQueueInterface, eventCount int) {
> ```

Why accept a `eventCount` from the caller, instead of just deciding inside this function? Does the caller care how many events are used in this shared testing helper?
___
*[`pkg/util/queue/queue_test.go` line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AwKsm1rgNStUDCVIT:-O4AwKsm1rgNStUDCVIU:b6tzeyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bf73d505537ccb3170df5ea5011f3fbffbba6d/pkg/util/queue/queue_test.go#L49)):*
> ```Go
> 	const eventCount = 2000000
> 
> 	q, err := NewQueue[*testQueueItem]()
> ```

The results of this benchmark are going to be skewed by the use of a pointer element type. That will incur a heap allocation per element, but not for the queue. I'd suggest switching to something basic like an `int64` for the purposes of the benchmark.
___
*[`pkg/util/queue/queue_test.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AwmQQ0aoRUmdPqZjP:-O4AwmQQ0aoRUmdPqZjQ:b-ni9i0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bf73d505537ccb3170df5ea5011f3fbffbba6d/pkg/util/queue/queue_test.go#L52)):*
> ```Go
> 	require.NoError(b, err)
> 
> 	for i := 0; i < b.N; i++ {
> ```

Do we need this first loop? The second loop seems to do everything the first does, and more. If there's a reason to differentiate them, should we pull them into different sub-benchmarks so that their performance can be understood in isolation?
___
*[`pkg/util/queue/queue_test.go` line 50 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AzmLU7geOJXbqR6Vq:-O4AzmLU7geOJXbqR6Vr:bs2eeq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/976fda2c5e8183d5d296296c09fa647c09327c38/pkg/util/queue/queue_test.go#L50)):*
> ```Go
> 	const eventCount = 2000000
> 
> 	testutils.RunTrueAndFalse(b, "queue with fixed chunk size", func(b *testing.B, fixedChunkSize bool) {
> ```

It's more common to see this kind of thing as a separate benchmark. So `BenchmarkQueueWithFixedChunkSize`.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128863)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`-- commits` line 37 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4Ax8WQDC3CEiHHpGYq):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

4000 elements is a strange number. Typically factors of 2 are used for data-structures, which has some benefits like fitting nicely in heap allocation size classes. Consider switching to 4096.
</blockquote></details>

Done.
___
*[`-- commits` line 48 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4B--p107BI32wPoriC):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

See my comment below about using a non-pointer type in the benchmark. These allocation numbers are hard to interpret when the queue's allocations are drowned out by the elements'.
</blockquote></details>

Ahh, makes more sense now. Updated the test to use int. 
___
*[`pkg/util/queue/queue.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4Av1pI7YBpwkyLvMk9:-O4BGoKO8u2jYHN522JQ:b-r8dzus) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdf073acdafc59b4abc61866e852b8f7d995e175/pkg/util/queue/queue.go#L102)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The following (`O(1)`) should have been enough to clear the queue and allow all chunks to be GCed. I wonder why that's not how this was written.
```go
q.head = nil
q.tail = nil
```
</blockquote></details>

I rewrote queue.removeAll to use this method. For QueueWithFixedChunkSize, we do need to iterate through the entire list and invoke the callback since each item will hold onto a shared memory allocation https://github.com/cockroachdb/cockroach/blob/f77445e9f5b45b398af3a9337c15215529eaa523/pkg/kv/kvserver/rangefeed/buffered_registration.go#L116 and should be properly released when removeAll is invoked. 
___
*[`pkg/util/queue/queue_with_fixed_chunk_size.go` line 65 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AyEV53A7n5GyUKgU0:-O4BILyH2hRZyZmjhkNy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/976fda2c5e8183d5d296296c09fa647c09327c38/pkg/util/queue/queue_with_fixed_chunk_size.go#L65)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why defer this? Can't we just `q.eventCount++` immediately?
</blockquote></details>

Done.
___
*[`pkg/util/queue/queue_with_fixed_chunk_size.go` line 89 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AycJK6WHT1ZssDcXk:-O4BIxvaCcb6kuIrO1HL:bda9sqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/976fda2c5e8183d5d296296c09fa647c09327c38/pkg/util/queue/queue_with_fixed_chunk_size.go#L89)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Isn't this just `q.Len() == 0`?
</blockquote></details>

Yeah, I think I just kept it the same as the current implementation of Queue. This should == `q.Len() == 0`. 
___
*[`pkg/util/queue/queue_with_fixed_chunk_size.go` line 117 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AzXvy1R8r6OpO2b3s:-O4BJAdI4GfJJJ53Xeu4:b4wemlq) ([raw file](https://github.com/cockroachdb/cockroach/blob/976fda2c5e8183d5d296296c09fa647c09327c38/pkg/util/queue/queue_with_fixed_chunk_size.go#L117)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need this callback? It causes the function to be less efficient and is currently required to be non-nil.
</blockquote></details>

See above. I was planning to use the callback to free the shared memory allocation held by each item. 
___
*[`pkg/util/queue/helpers_test.go` line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AvFOK2Yu7QtdBXmMC:-O4BJLPT5_7jBOSaSYpm:b458l0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bf73d505537ccb3170df5ea5011f3fbffbba6d/pkg/util/queue/helpers_test.go#L31)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What is the int64 param?
</blockquote></details>

I refactored the code to remove this usage by adding another checkEventCount method. 
___
*[`pkg/util/queue/helpers_test.go` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AuUJ0052UJOUcYOdm:-O4BJTdO-pCp6GaIzaut:ba0bhkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bf73d505537ccb3170df5ea5011f3fbffbba6d/pkg/util/queue/helpers_test.go#L35)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Has this file (at this commit) been run through gofmt?
</blockquote></details>

Hm, I thought I have built-in gofmt. Is there something you feel non-idiomatic?
___
*[`pkg/util/queue/helpers_test.go` line 78 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4Aw-l-7y2eBWi61WNV:-O4BK0DO9EtDmWS_xaT4:b-6p60a4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bf73d505537ccb3170df5ea5011f3fbffbba6d/pkg/util/queue/helpers_test.go#L78)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Am I missing something, or are we not using this test with `queueChunkWithFixedSize`?
</blockquote></details>

Oops, no wonder why tests passed on the first run. 
___
*[`pkg/util/queue/helpers_test.go` line 78 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4Aw64t07Ea4MSD2bal:-O4BK4PM-H13Zt1WymRP:b-9sxsp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bf73d505537ccb3170df5ea5011f3fbffbba6d/pkg/util/queue/helpers_test.go#L78)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why accept a `eventCount` from the caller, instead of just deciding inside this function? Does the caller care how many events are used in this shared testing helper?
</blockquote></details>

You are right. The caller shouldn't care. I've cleaned it up. 
___
*[`pkg/util/queue/queue_test.go` line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AwKsm1rgNStUDCVIT:-O4BKGj_CEeHt9I_S4F9:bxzxivj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bf73d505537ccb3170df5ea5011f3fbffbba6d/pkg/util/queue/queue_test.go#L49)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The results of this benchmark are going to be skewed by the use of a pointer element type. That will incur a heap allocation per element, but not for the queue. I'd suggest switching to something basic like an `int64` for the purposes of the benchmark.
</blockquote></details>

Ah, makes more sense now. This is the new benchmark result. 

```
BenchmarkQueueWithFixedChunkSize
BenchmarkQueueWithFixedChunkSize           18880            116079 ns/op               9 B/op          0 allocs/op
BenchmarkQueue
BenchmarkQueue                             28417            121969 ns/op          239060 B/op        446 allocs/op
```
___
*[`pkg/util/queue/queue_test.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AwmQQ0aoRUmdPqZjP:-O4BKOJ4DiStWNc4NB2V:b-ym8wfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bf73d505537ccb3170df5ea5011f3fbffbba6d/pkg/util/queue/queue_test.go#L52)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need this first loop? The second loop seems to do everything the first does, and more. If there's a reason to differentiate them, should we pull them into different sub-benchmarks so that their performance can be understood in isolation?
</blockquote></details>

I changed them a bit. One is doing a mixed of Enqueue and Dequeue. The other is calling Enqueue + removeAll. 
___
*[`pkg/util/queue/queue_test.go` line 50 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AzmLU7geOJXbqR6Vq:-O4BKNakELmSTpt_3jkH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/976fda2c5e8183d5d296296c09fa647c09327c38/pkg/util/queue/queue_test.go#L50)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's more common to see this kind of thing as a separate benchmark. So `BenchmarkQueueWithFixedChunkSize`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
Apologies for so many force pushes :grimacing: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-:-O4BQgRe9UeJyTaCpQzD:b30k95m)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/util/queue/queue.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4Av1pI7YBpwkyLvMk9:-O4BQV-oDkfUWgMuO3-u:b-swwi9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdf073acdafc59b4abc61866e852b8f7d995e175/pkg/util/queue/queue.go#L102)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

I rewrote queue.removeAll to use this method. For QueueWithFixedChunkSize, we do need to iterate through the entire list and invoke the callback since each item will hold onto a shared memory allocation https://github.com/cockroachdb/cockroach/blob/f77445e9f5b45b398af3a9337c15215529eaa523/pkg/kv/kvserver/rangefeed/buffered_registration.go#L116 and should be properly released when removeAll is invoked.
</blockquote></details>

Nvm, I dropped the changes for queue.removeAll since it doesn't give us an apple to apple comparison for the benchmark. queue.removeAll is only used for testing though. 

```
BenchmarkQueueWithFixedChunkSize
BenchmarkQueueWithFixedChunkSize           26445            118984 ns/op              11 B/op          0 allocs/op
BenchmarkQueue
BenchmarkQueue                             28495            124318 ns/op          239060 B/op        446 allocs/op
PASS
```


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/util/queue/helpers_test.go` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128863#-O4AuUJ0052UJOUcYOdm:-O4CKj8m7eshlBqG3U_M:b-r94l7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bf73d505537ccb3170df5ea5011f3fbffbba6d/pkg/util/queue/helpers_test.go#L35)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Hm, I thought I have built-in gofmt. Is there something you feel non-idiomatic?
</blockquote></details>

I think I see why you said this. Were you referring to having struct receiver method in a test file? Linter complained about it. I've changed it.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
util/queue: fix linter failure

TestLint/TestStaticCheck has some issues with generic methods. This patch works
around the issue by disabling them for these functions.

Epic: none
Release note: none
